Added support for picking a given build from the reference job

eivindt opened the pull request:
We automatically create a Jenkins job per pull request, and want to compare various statistics with the version at the fork point.  In order to have the necessary data available for the fork points, we have a separate 'baseline' job that we can tell to build a certain revision.

When we found the _referenceJob_ parameter in this plugin, this seemed like a perfect match, but since we have many pull requests, the baseline job may contain many builds of fork points, and picking the latest will not work.

This pull request adds support for another reference parameter, _referenceBuild_, letting you pick a certain build to use for the reference.  We use it in a pipeline, like this:

```
baseline_build = selectRun filter: parameters("commit=${FORK_POINT}"), job:"__baselines__"
recordIssues enabledForFailure: true, tool: cppCheck(pattern: 'build/ssca.xml'), referenceJobName: '__baselines__', referenceBuildId: baseline_build.getId()
```

As Jenkins plugins always confuse me, there's a chance I've done something wrong or stupid, but it seems to work as intended.

I did not add form validation for the build id, and I did not update the documentation (I was not able to create new screenshots using the same theme/os as the originals).

Let me know if there is more I should do in order to get this accepted.

uhafner commented:
Thank you for creating this new feature! Your pull request addresses an important missing feature of the plugin. 

Currently one of my students is extending the plugin to create a general solution for that problem: we try to automatically identify the branch point of a PR (see [JENKINS-33274](https://issues.jenkins-ci.org/browse/JENKINS-33274))and therefore the correct reference point for the analysis. I think I need to discuss your PR in more detail with him: I'm not sure if it makes sense to change the API now even though we will get a more general solution in the future.

eivindt commented:
Thanks for the feedback. A more general solution is probably a good idea, and may make more sense for your plugin.

For our use, we prefer being able to reference a build in another job. It would be nice if Jenkins offered an easier way to select a fork point build, the way we do it today is quite messy and involves a combination of shell scripts and groovy code.

Our reference job provides more information than I think the warnings plugin is capable of providing, like image build sizes for various build parameters, available stack size and runtime time measurements.

Some of these are quite expensive (time wise), which is why we prefer to (re)use a reference build's data for these measurements (including those that warnings-ng is able to give us a really nice interface for viewing).

Anyway, that's just a bit more information on how we intend to use this.

uhafner commented on src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesRecorder.java:
```diff
@@ -435,6 +437,34 @@ public String getReferenceJobName() {
+    public void setReferenceBuildId(final String referenceBuildId) {
+        if (NO_REFERENCE_BUILD.equals(referenceBuildId)) {
+            this.referenceBuildId = StringUtils.EMPTY;
+        } else {
```
newline before else or remove the else as it is redundant anyway

eivindt replied:
Newline before else added.

uhafner commented on src/main/java/io/jenkins/plugins/analysis/core/steps/RecordIssuesStep.java:
```diff
@@ -837,6 +838,34 @@ public String getReferenceJobName() {
+    public void setReferenceBuildId(final String referenceBuildId) {
+        if (IssuesRecorder.NO_REFERENCE_BUILD.equals(referenceBuildId)) {
+            this.referenceBuildId = StringUtils.EMPTY;
+        } else {
```
see above

uhafner commented on src/main/java/io/jenkins/plugins/analysis/core/steps/RecordIssuesStep.java:
```diff
@@ -68,6 +68,7 @@
     private boolean ignoreQualityGate = false; // by default, a successful quality gate is mandatory;
     private boolean ignoreFailedBuilds = true; // by default, failed builds are ignored;
     private String referenceJobName;
+    private String referenceBuildId;
```
Initialize with ` = StringUtils.EMPTY` (and add the same thing for line 70 since otherwise the snipped generator will generate an empty statement even if the string is empty.

eivindt replied:
Done

uhafner commented on src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesRecorder.java:
```diff
@@ -435,6 +437,34 @@ public String getReferenceJobName() {
+     * @param referenceBuildId
+     *         the build id of the reference job
+     */
+    @DataBoundSetter
```
Remove `@DataBoundSetter`: I'm not sure if this is correct but I think that a Freestyle Build will never be configured in that way, your new option makes sense only for pipelines where you are computing the build ID on your own.

eivindt replied:
I only added this because all other configuration options had it.  It probably does not make sense outside of a pipeline, at least unless support is added for a more dynamic UI based build id selector.  
I've removed this.

uhafner commented on src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesRecorder.java:
```diff
@@ -1131,6 +1161,10 @@ public FormValidation doCheckReferenceJobName(@QueryParameter final String refer
             return model.validateJob(referenceJobName);
         }
 
+        public FormValidation doCheckReferenceBuildId(@QueryParameter final String referenceBuildId) {
```
remove this method as this field cannot be part of the validation (you are returning true in any case)

uhafner commented on src/main/java/io/jenkins/plugins/analysis/core/util/ModelValidation.java:
```diff
@@ -31,6 +31,8 @@
 
     @VisibleForTesting
     static final String NO_REFERENCE_JOB = "-";
+    @VisibleForTesting
```
This file needs no modifications (since not meaningful for freestyle jobs)

eivindt replied:
Fixed.

uhafner commented on src/main/java/io/jenkins/plugins/analysis/core/util/ModelValidation.java:
```diff
@@ -151,6 +153,10 @@ public FormValidation validateJob(final String referenceJobName) {
         return FormValidation.error(Messages.FieldValidator_Error_ReferenceJobDoesNotExist());
     }
 
+    public FormValidation validateBuild(final String referenceBuildId) {
```
see above

uhafner commented on src/main/resources/issues/publish-parameters.jelly:
```diff
@@ -27,6 +27,9 @@
   <f:entry title="${%title.referenceJob}" field="referenceJobName" description="${%description.referenceJob}">
     <f:combobox/>
   </f:entry>
+  <f:entry title="${%title.referenceBuildId}" field="referenceBuildId" description="${%description.referenceBuildId}">
```
I think a UI configuration makes no sense, or do you use the snippet generator on your own for this type of configuration? So it would be save to remove these lines.

eivindt replied:
Removed.

uhafner commented on src/main/resources/issues/publish-parameters.properties:
```diff
@@ -3,6 +3,8 @@ description.reference=One important feature of the Warnings Plugin is the classi
   New, fixed, and outstanding issues are then computed by comparing the issues in the current build and in the baseline.
 title.referenceJob=Reference Job
 description.referenceJob=The reference job provides the builds that can be selected as baseline.
+title.referenceBuildId=Reference Build Id
```
see above

uhafner commented on src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesPublisher.java:
```diff
@@ -198,7 +200,11 @@ private History createAnalysisHistory(final ResultSelector selector, final Repor
-                baseline = referenceJob.get().getLastBuild();
+                if (referenceBuildId != null) {
+                    baseline = referenceJob.get().getBuild(referenceBuildId);
+                } else {
```
newline before else

uhafner commented on src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesPublisher.java:
```diff
@@ -198,7 +200,11 @@ private History createAnalysisHistory(final ResultSelector selector, final Repor
+                } else {
+                    baseline = referenceJob.get().getLastBuild();
+                }
             }
```
I think I have overseen a bug in my code (that is now in your code as well): the baseline can be `null` (build has been deleted, etc.). So it would make sense to add after the if-else (i.e. for both cases) a 
```
if (baseline == null) {
   return new NullHistory();
}
```

It would also make sense to test that in an additional test.

eivindt replied:
I've added the check, and made two more integration tests:

- use a reference job with no builds
- use a reference job with builds, but referring to a non existent build id

I was sort of expecting these tests to show all issues as "new" since the baseline was missing, but they are "outstanding".  I assume this is intentional.

I *have* had this issue on my local install, and this has caused build failures.  I am slightly worried that with this change, a missing baseline build may not show up as an error at all, and will therefore be hidden.

uhafner replied:
The first build of a job cannot create new warnings, this is intentional, yes.

But is is a good point if this should be ignored or produce an error. Maybe a better way is to use something like 

```
if (referenceBuildId == null) {
   baseline = referenceJob.get().getLastBuild();
   if (baseline == null) {
      throw new IllegalArgumentException(String.format("Reference job '%s' does not contain a valid build", referenceJob);
   }
}
else {
   baseline = referenceJob.get().getBuild(referenceBuildId);
   if (baseline == null) {
      throw new IllegalArgumentException(String.format("Reference job '%s' does not contain build '%s'", 
referenceJob, referenceBuildId );
   }
}
```  
What do you think? Or we can create an error message using `filtered.logError`. That will be flagged as an error triangle at the build page.

eivindt replied:
In this case it's the first build of job that has explicitly been told to compare warnings with another job or a specific build of another job.

If the specific build does not exist, I'd say that's an error.  If you only specified a reference job, and that job does not exist, I'd say that's an error.  If you specify a reference job, and that job does not have any builds to compare warnings to, I'd still say that's an error -- that configuration does not make sense.

I think your suggestion looks good, let me know, and I'll add it to the pull request.

uhafner replied:
What kind of error do you prefer? Using the logger.error method will create something like https://ci.jenkins.io/job/Plugins/job/warnings-ng-plugin/job/reference/1/
(Open Tasks error triangle - that you can click and see the details of the error messages). Or an exception that fails the build? This seems a little bit too hard for me and is not visualized in the build...

eivindt replied:
That's a good job for illustration.

I agree, failing the build would be too hard, the error triangle is the best option.

I hope to amend the pull request tomorrow.

uhafner replied:
Thank you very much!

uhafner commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/uhafner/warnings-ng-plugin/pullRequest?prid=4335332&bid=14743939)

uhafner commented on src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesRecorder.java:
```diff
@@ -435,6 +437,34 @@ public String getReferenceJobName() {
+        if (NO_REFERENCE_BUILD.equals(referenceBuildId)) {
+            this.referenceBuildId = StringUtils.EMPTY;
+        }
+        else {
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['}' at column 9 should be alone on a line.](https://app.codacy.com/manual/uhafner/warnings-ng-plugin/pullRequest?prid=4335332)

uhafner replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['}' at column 9 should be alone on a line.](https://app.codacy.com/manual/uhafner/warnings-ng-plugin/pullRequest?prid=4335332)

uhafner replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['}' at column 9 should be alone on a line.](https://app.codacy.com/manual/uhafner/warnings-ng-plugin/pullRequest?prid=4335332)

uhafner commented on src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesPublisher.java:
```diff
@@ -198,7 +201,15 @@ private History createAnalysisHistory(final ResultSelector selector, final Repor
             Optional<Job<?, ?>> referenceJob = new JenkinsFacade().getJob(referenceJobName);
             if (referenceJob.isPresent()) {
-                baseline = referenceJob.get().getLastBuild();
+                if (referenceBuildId != null) {
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid if (x != y) ..; else ..;](https://app.codacy.com/manual/uhafner/warnings-ng-plugin/pullRequest?prid=4335332)

uhafner commented on src/main/java/io/jenkins/plugins/analysis/core/steps/RecordIssuesStep.java:
```diff
@@ -837,6 +838,35 @@ public String getReferenceJobName() {
+    public void setReferenceBuildId(final String referenceBuildId) {
+        if (IssuesRecorder.NO_REFERENCE_BUILD.equals(referenceBuildId)) {
+            this.referenceBuildId = StringUtils.EMPTY;
+        }
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['}' at column 9 should be alone on a line.](https://app.codacy.com/manual/uhafner/warnings-ng-plugin/pullRequest?prid=4335332)

uhafner replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['}' at column 9 should be alone on a line.](https://app.codacy.com/manual/uhafner/warnings-ng-plugin/pullRequest?prid=4335332)

uhafner replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['}' at column 9 should be alone on a line.](https://app.codacy.com/manual/uhafner/warnings-ng-plugin/pullRequest?prid=4335332)

uhafner commented on src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesPublisher.java:
```diff
@@ -198,7 +201,15 @@ private History createAnalysisHistory(final ResultSelector selector, final Repor
             if (referenceJob.isPresent()) {
-                baseline = referenceJob.get().getLastBuild();
+                if (referenceBuildId != null) {
+                    baseline = referenceJob.get().getBuild(referenceBuildId);
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['}' at column 17 should be alone on a line.](https://app.codacy.com/manual/uhafner/warnings-ng-plugin/pullRequest?prid=4335332)

uhafner replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['}' at column 17 should be alone on a line.](https://app.codacy.com/manual/uhafner/warnings-ng-plugin/pullRequest?prid=4335332)

uhafner commented on src/main/java/io/jenkins/plugins/analysis/core/steps/RecordIssuesStep.java:
```diff
@@ -837,6 +838,35 @@ public String getReferenceJobName() {
+        if (IssuesRecorder.NO_REFERENCE_BUILD.equals(referenceBuildId)) {
+            this.referenceBuildId = StringUtils.EMPTY;
+        }
+        else {
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [File contains tab characters (this is the first instance).](https://app.codacy.com/manual/uhafner/warnings-ng-plugin/pullRequest?prid=4335332)

eivindt replied:
fixed

uhafner commented on src/test/java/io/jenkins/plugins/analysis/warnings/StepsITest.java:
```diff
@@ -685,6 +685,89 @@ public void shouldUseOtherJobAsReference() {
+        WorkflowJob reference = createPipeline("reference");
+        copyMultipleFilesToWorkspaceWithSuffix(reference, "java-start-rev0.txt");
+        reference.setDefinition(createPipelineScriptWithScanAndPublishSteps(new Java()));
+        AnalysisResult firstReferenceResult = scheduleSuccessfulBuild(reference);
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid unused local variables such as 'firstReferenceResult'.](https://app.codacy.com/manual/uhafner/warnings-ng-plugin/pullRequest?prid=4335332)

eivindt replied:
fixed

uhafner commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 5
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/uhafner/warnings-ng-plugin/pullRequest?prid=4335332&bid=14743939)

uhafner commented on src/main/java/io/jenkins/plugins/analysis/core/steps/RecordIssuesStep.java:
```diff
@@ -837,6 +838,35 @@ public String getReferenceJobName() {
+        if (IssuesRecorder.NO_REFERENCE_BUILD.equals(referenceBuildId)) {
+            this.referenceBuildId = StringUtils.EMPTY;
+        }
+        else {
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [File contains tab characters (this is the first instance).](https://app.codacy.com/manual/uhafner/warnings-ng-plugin/pullRequest?prid=4335332)

uhafner commented on src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesPublisher.java:
```diff
@@ -198,7 +201,15 @@ private History createAnalysisHistory(final ResultSelector selector, final Repor
             Optional<Job<?, ?>> referenceJob = new JenkinsFacade().getJob(referenceJobName);
             if (referenceJob.isPresent()) {
-                baseline = referenceJob.get().getLastBuild();
+                if (referenceBuildId != null) {
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid if (x != y) ..; else ..;](https://app.codacy.com/manual/uhafner/warnings-ng-plugin/pullRequest?prid=4335332)

eivindt replied:
rephrased

uhafner commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/uhafner/warnings-ng-plugin/pullRequest?prid=4335332&bid=14743939)

uhafner commented on src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesPublisher.java:
```diff
@@ -198,7 +201,15 @@ private History createAnalysisHistory(final ResultSelector selector, final Repor
             if (referenceJob.isPresent()) {
-                baseline = referenceJob.get().getLastBuild();
+                if (referenceBuildId == null) {
+                    baseline = referenceJob.get().getLastBuild();
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['}' at column 17 should be alone on a line.](https://app.codacy.com/manual/uhafner/warnings-ng-plugin/pullRequest?prid=4335332)

uhafner commented on src/main/java/io/jenkins/plugins/analysis/core/steps/RecordIssuesStep.java:
```diff
@@ -837,6 +838,35 @@ public String getReferenceJobName() {
+        if (IssuesRecorder.NO_REFERENCE_BUILD.equals(referenceBuildId)) {
+            this.referenceBuildId = StringUtils.EMPTY;
+        }
+        else {
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [File contains tab characters (this is the first instance).](https://app.codacy.com/manual/uhafner/warnings-ng-plugin/pullRequest?prid=4335332)

eivindt replied:
This is fixed, but the codacy tool doesn't seem to realise the file has changed.

eivindt commented:
I've updated the pull request, handling missing reference builds as logged errors, rather than failing the build.

I've also rebased it against 7.3.0.

uhafner commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesPublisher.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/uhafner/warnings-ng-plugin/pullRequest?prid=4335332&bid=14743939)

uhafner commented:
Thank you very much!
